arrows/core/csv_io.h

```diff
@@ -0,0 +1,359 @@
+struct KWIVER_ALGO_CORE_EXPORT comment_t {};
+
+KWIVER_ALGO_CORE_EXPORT
+extern comment_t const comment;
+
+} // namespace csv
+
+// ----------------------------------------------------------------------------
+/// A barebones yet flexible CSV writer.
+///
+/// This class abstracts away writing delimiters and properly quoting fields.
+/// No assumptions are made about consistent row lengths or column data types.
+/// A newline is automatically added at the end of the file if there is not one
+/// already.
+///
+/// \todo Add a way to specify precision and fixed/scientific for float values.
+/// \todo Add an unprocessed string write option, to facilitate writing
+///       comments that are not meant to mimic data fields.
+class KWIVER_ALGO_CORE_EXPORT csv_writer
+{
+public:
+  /// Construct a csv writer.
+  ///
+  /// \param os Output stream to write to.
+  /// \param delim Character that separates fields.
+  /// \param quote Character that begins and ends a quoted field.
+  /// \param quote_esc
+  ///   Character that must be used to escape \p quote and itself when they
+  ///   appear in quoted fields. May be the same as \p quote.
+  /// \param comment
+  ///   Character that, when appearing as the first character in a line,
+  ///   designates that line as a comment.
+  csv_writer(
+    std::ostream& os,
+    char delim = csv::DEFAULT_DELIM,
+    char quote = csv::DEFAULT_QUOTE,
+    char quote_esc = csv::DEFAULT_QUOTE_ESC,
+    char comment = csv::DEFAULT_COMMENT );
+  ~csv_writer();
+
+  /// Write a field of type \c T to the file, or perform a special operation.
+  ///
+  /// Passing \c csv::begf begins a field, causing future writes to concatenate
+  /// their \p value together instead of writing them as separate fields, until
+  /// \c csv::endf is passed. Passing \c csv::skipf writes an empty field.
+  /// Passing \c csv::endl ends the current line. Passing \c csv::comment
+  /// starts a comment on the current line. Writing while in a comment behaves
+  /// in the exact same way as writing outside a comment, including quoting and
+  /// delimiting behaviors.
+  ///
+  /// \throws std::invalid_argument If the operation is not valid.
+  template < class T >
+  csv_writer& write( T const& value );
+
+  /// Returns the delimiting character in use.
+  char delim() const;
+
+  /// Returns the quote character in use.
+  char quote() const;
+
+  /// Returns the quote escape character in use.
+  char quote_esc() const;
+
+  /// Returns the comment character in use.
+  char comment() const;
+
+private:
+  void commit();
+
+  std::ostream& m_os;
+  std::stringstream m_ss;
+  char m_delim;
+  char m_quote;
+  char m_quote_esc;
+  char m_comment;
+  bool m_in_field;
+  bool m_first_field;
+};
+
+// ----------------------------------------------------------------------------
+/// Write \p value as a CSV field.
+KWIVER_ALGO_CORE_EXPORT
+csv_writer& operator<<( csv_writer& os, char const* value );
```

Comment:
Unfortunately not, since the type of the literal `"string"` is not `char const*` but `char [7]` or something. So you end up with an unbounded number of different template instantiations based on the length of the string literal.